### PR TITLE
fix(shell-integration): emit OSC 133 C on every accepted line and stop deleting the user's ssh

### PR DESCRIPTION
**Pull request by amanthanvi**
## Summary

Two independent defects in `src/shell-integration/powershell/integration.ps1`.

**OSC 133 C never fired for most users, and clobbered the user's handler.** It
was hooked on `Set-PSReadLineOption -CommandValidationHandler`. PSReadLine
invokes that handler from exactly one place, its `ValidateAndAcceptLine`
function, and the default Enter binding on both pwsh 7 and Windows PowerShell
5.1 is `AcceptLine` -- so the mark was never emitted for any user who had not
rebound Enter, which left *copy last command output* and *insert last command*
dead on PowerShell. It also replaced, without chaining to, any
`CommandValidationHandler` the user's profile had installed. It now rides on
`AddToHistoryHandler`, which PSReadLine calls for every accepted line whatever
Enter is bound to, and calls *before* the host executes the line. The old hook
is removed rather than kept, so users who *do* bind `ValidateAndAcceptLine` get
one C mark, not two.

**A user's own `ssh` function was deleted on every launch.** The `ssh` block
opened with an unconditional
`Remove-Item -Path Function:\ssh,Function:\global:ssh -ErrorAction SilentlyContinue`
placed *ahead* of its own `ssh-*` feature check, so a `function ssh` defined in
the user's `$PROFILE` -- which runs before noctty's injected `-Command` -- was
destroyed even when `GHOSTTY_SHELL_FEATURES` carried no `ssh-*` feature and
nothing replaced it. `Function:\global:ssh` is also not a valid provider path
(a scope qualifier belongs on the *name*, not inside a `Function:` drive path),
so the same line raised `ItemNotFoundException` every launch. Removal is now
gated on a marker in the function body plus a session flag that cannot be set
on a first load, so only a wrapper this script installed is ever removed.

Everything below was measured on this machine on **pwsh 7.6.6 / PSReadLine
2.4.5** and **Windows PowerShell 5.1.26100 / PSReadLine 2.0.0**, not inferred:

- **`AddToHistoryHandler` really does fire under the default `AcceptLine`
  binding on both hosts, ~8 ms before the command runs.** Verified by driving
  genuinely interactive sessions via `FreeConsole` then `AttachConsole(pid)`
  then `WriteConsoleInputW`. Piping stdin into a `-NoExit -Command` child does
  not exercise PSReadLine at all, and `SendKeys` cannot reach a host launched
  under Windows Terminal because the spawned process owns no window
  (`MainWindowHandle=0`). `CommandValidationHandler` was confirmed *not* to
  fire under `AcceptLine` on either host, and to start firing in the same
  process once Enter was rebound to `ValidateAndAcceptLine` -- so that negative
  is a real negative, not a broken probe.
- **The option is typed `Func[string, object]`, not `[scriptblock]`.**
  PowerShell converts at parameter-bind time, so reading the option back yields
  a delegate whose `ToString()` is the bare delegate type name and whose
  `Target` is a compiler-generated `Closure` exposing only `Constants` /
  `Locals`. There is no script text left to carry a marker, so "is the
  installed handler ours?" is answered by reference equality against the
  delegate read back after installing (measured stable across reads). This is a
  deliberate deviation from the "marker in its script text" approach -- that is
  not available for this property.
- **The predecessor handler is never null.** Both PSReadLine versions ship a
  default that returns `MemoryOnly` for credential-shaped lines to keep them
  out of the on-disk history file. Chaining is therefore a security
  requirement, not a courtesy: we call it and return its answer unchanged. Our
  handler's own output is kept a scalar, because a scriptblock-derived `Func`
  returns the block's whole output collection and one stray pipeline value
  would turn a `$true` into an `object[]` PSReadLine cannot interpret.
- **PSReadLine 2.0.0 (5.1 only) also calls the handler for every line it
  replays out of the shared `ConsoleHost_history.txt`** -- 1790 invocations
  before a single keypress on a fresh session -- and for lines other live
  sessions append. Unhandled, that is a storm of bogus command marks at
  startup. Emission is gated on
  `[Microsoft.PowerShell.PSConsoleReadLine]::GetBufferState` *equalling* the
  line (`-ceq`). A `cursor -gt 0` test is not enough: a foreign line arrives
  while the user's own text is in the buffer. A prompt-armed one-shot does not
  work either -- the replay burst happens after the first prompt draw and eats
  it. The gate fails closed, and the chained handler still runs for every
  invocation, so suppressing the terminal report never changes the history
  decision.
- **`-ErrorAction SilentlyContinue` suppresses the message but still appends
  the record to `$Error`.** Measured `$Error.Count` 0 to 2 for the old
  `Remove-Item` line on both hosts. The probe lookups (`Get-Variable` for the
  launch variable and for `$LASTEXITCODE`, `Get-Command` for `ssh` and for the
  ssh application) now use `-ErrorAction Ignore`, which leaves `$Error.Count`
  at 0 and still yields `$null` under `Set-StrictMode -Version Latest`. The
  `$LASTEXITCODE` one mattered beyond load: it runs several times per prompt
  draw.
- Growing the embedded script exhausted `@setEvalBranchQuota(1_000_000)` on the
  comptime `Sha256` of `integration_script`. The failure surfaces inside
  `std/crypto/sha2.zig` with nothing pointing at the `.ps1`; the quota scales
  with script length and is now `10_000_000`.

Docs updated in `docs/windows-capability-matrix.md` and
`src/shell-integration/README.md`, plus the doc comment in `integration.ps1`
and the harness comment.

One correction to the brief this was written from: *insert last command* and
*copy last command output* are driven purely by OSC 133 marks in this codebase
-- there is no `Get-History` path (`grep -r "Get-History"` returns nothing).
The docs say they now work by default on PowerShell because the C mark is now
emitted, without claiming a `Get-History` mechanism that does not exist.

## Validation

- [x] `test/windows/powershell-shell-integration.ps1` prints `PASS powershell
      shell integration` under **both** `pwsh -NoProfile -File ...` and
      `powershell -NoProfile -File ...`.
- [x] `zig build test -Dtest-filter=powershell`
- [x] `zig build test -Dtest-filter=ConPTY -Dtest-filter=powershell`
- [x] `zig build test -Dtest-filter=ConPTY -Dtest-filter=honours`
- [x] `zig build -Demit-exe=true`
- [x] `zig fmt` clean on the touched files
- [x] `AGENTS.md` blob is CR-free (carriage-return count on `HEAD:AGENTS.md` is 0)

New harness cases:

- a user-defined `function global:ssh` survives a launch with no `ssh-*`
  feature, and load leaves `$Error.Count` at 0 -- both asserted in a real child
  process running the real injected block argv, not in-process;
- the wrapper *is* installed and carries its ownership marker when `ssh-env` is
  on, again in a real child, also with `$Error.Count` 0;
- re-sourcing replaces our wrapper without stacking a second one;
- a pre-installed `AddToHistoryHandler` is chained, receives the exact line,
  and its `AddToHistoryOption` answer is returned unchanged and unreshaped;
- a replayed history line (empty PSReadLine buffer) reaches the chained handler
  but emits no `133;C`, and the discriminator itself is asserted directly;
- re-sourcing produces exactly one `133;C` per line.

Scope-guard counts move to 17 declarations / 9 top-level variables. Its
substring assertions now scan code lines only, via a new `codeContains` helper,
because the script's comments quote the very constructs the tests assert are
gone; a companion test keeps trailing `#` comments off code lines so that
helper stays sound.

## Risks / Follow-ups

- The accepted-line path in the harness stubs the discriminator, because a
  non-interactive host cannot put text in PSReadLine's buffer
  (`PSConsoleReadLine::Insert` throws `NullReferenceException` with no console
  attached, on both hosts). The discriminator itself is asserted directly, and
  the end-to-end "fires under `AcceptLine`" behaviour was verified by console
  input injection, but that end-to-end path is not covered by an automated
  test.
- Residual edge case: if another live session appends a history line
  byte-identical to what the user just typed, 5.1 emits one extra `133;C`.
  Rare, and one spurious mark rather than a storm.
- If something replaces our `AddToHistoryHandler` with a wrapper *around* it
  and the script is then re-sourced, we chain to that wrapper and the mark
  would be emitted twice. Plain re-sourcing is idempotent; this shape is not
  handled.
- `-ErrorAction Ignore` records cannot be recovered from `$Error` for
  debugging. That is the point at these probe sites, but it is a behaviour
  change from `SilentlyContinue`.
- The comptime SHA-256 branch quota will need raising again if the script grows
  several times over; the failure mode is confusing enough that a comment now
  says so at the call site.

## AI assistance

Per AI_POLICY.md: investigated, implemented and validated by an AI agent
(Claude Code) operating the maintainer's account with the maintainer's
authorization.

## Summary by Sourcery

Fix PowerShell command markers and ssh integration so accepted commands are reported reliably without overriding history policies or user-defined functions.

Bug Fixes:
- Emit OSC 133 C for every accepted PowerShell line by using PSReadLine's AddToHistoryHandler, restoring command output copying and last-command insertion by default.
- Preserve and chain existing PSReadLine history handlers, including protection for sensitive history entries and suppression of marks for replayed history lines.
- Stop deleting user-defined ssh functions and remove only noctty-owned wrappers, while eliminating launch-time PowerShell error records.

Enhancements:
- Make PowerShell shell integration idempotent across re-sourcing and ensure accepted-line handling returns the user's history policy unchanged.
- Expand Windows PowerShell integration coverage for handler chaining, history replay, accepted commands, ssh ownership, and error cleanliness.

Build:
- Increase the comptime evaluation quota used to hash the embedded PowerShell integration script.

Documentation:
- Update PowerShell capability and shell integration documentation to describe default OSC 133 C emission, handler preservation, replay filtering, and safe ssh wrapping.

Tests:
- Add validation for AddToHistoryHandler behavior, re-sourcing, replay suppression, chained results, failure-closed handling, and ssh wrapper ownership in both PowerShell hosts.

Chores:
- Record the PowerShell integration findings and corrective guidance in the self-correction log.

## Summary by CodeRabbit

* **New Features**
  * PowerShell integration now emits command-start markers reliably for accepted commands, including default Enter behavior.
  * Command-output copying and recoverable-command insertion now work by default in PowerShell.
  * Existing command-history behavior, including credential protection, is preserved.

* **Bug Fixes**
  * Replayed or externally added history entries no longer emit duplicate command markers.
  * User-defined PowerShell `ssh` functions are preserved when no integration wrapper is installed.
  * Reduced spurious PowerShell error-history entries during shell integration.

### CODE REVIEW

File: src/shell-integration/powershell/integration.ps1 [hunk 1]
```diff
@@ -425,34 +479,140 @@ function global:prompt {
...
+
+    $previous = __ghostty_read_global '__ghostty_addtohistory_original'
+    if ($null -eq $previous) { return $true }
+    try {
+        if ($previous -is [scriptblock]) { return (& $previous $Line) }
+        return $previous.Invoke($Line)
+    } catch {
+        # A predecessor that throws must not cost the user their history
+        # either. `$true` is PSReadLine's own "add it" answer.
+        return $true
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/shell-integration/powershell/integration.ps1]: Valid, and fixed in 84dd2a5.

You are right about the asymmetry: unless a profile replaced it, the predecessor *is* PSReadLine's sensitive-history scrubber (non-null out of the box on both PSReadLine 2.4.5 and 2.0.0, returning `MemoryOnly` for credential-shaped lines), so inventing `$true` when it throws is exactly the wrong direction — it persists to `ConsoleHost_history.txt` a line the predecessor may have been in the middle of holding back.

The catch now returns `[Microsoft.PowerShell.AddToHistoryOption]::MemoryOnly`, which keeps the line usable in the session (the user still gets it on Up-arrow) without writing it to disk. `SkipAdding` would fail closed too, but it silently costs the user their history entry for a failure that is ours, not theirs; `MemoryOnly` is the same answer PSReadLine's own default gives a line it distrusts.

One implementation detail worth flagging: the enum is resolved with `'Microsoft.PowerShell.AddToHistoryOption' -as [type]` rather than a `[...]` type literal, because an unresolvable type literal throws, and throwing from inside this catch would escape the handler entirely. If the type is absent the fallback is `$false`, PSReadLine's documented equivalent of `SkipAdding`.

The sibling `if ($null -eq $previous) { return $true }` is deliberately left as-is: that is the genuinely-no-predecessor case, where there is no policy to preserve and `MemoryAndFile` is what PSReadLine does on its own.

Covered by a new harness case that installs a predecessor which throws and asserts both that the result is `MemoryOnly` and that the OSC 133 C mark is still emitted. `PASS powershell shell integration` under both `pwsh 7.6.6` and `powershell.exe 5.1.26100`; `zig build test -Dtest-filter=ConPTY -Dtest-filter=powershell` and `-Dtest-filter=ConPTY -Dtest-filter=honours` both exit 0.

---
*Reply written by an AI agent (Claude Code) operating the maintainer's account, with the maintainer's authorization.*
